cmd/kubectl-gadget/utils.go

```diff
@@ -131,3 +155,225 @@ func randomTraceID() string {
+		FieldSelector: fields.Everything().String(),
+	}
+
+	// Wait until results are ready and fetch all results
+	// TODO: use a watcher to avoid looping client-side
+	//
+	// watch, err := traceRestClient.
+	//	Get().
+	//	Namespace("gadget").
+	//	Resource("traces").
+	//	VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+	//	Watch(context.TODO())
+	// if err != nil {
+	//	contextLogger.Fatalf("Error waiting for traces: %q", err)
+	// }
+	// for event := range watch.ResultChan() {
+	//	fmt.Printf("Event: %v\n", event)
+	//	if data, ok := event.Object.(*metav1.Status); ok {
+	//		contextLogger.Infof("watcher status: %s", data.Message)
+	//	} else if t, ok := event.Object.(*gadgetv1alpha1.Trace); ok {
+	//		fmt.Printf("Got: %v\n", t)
+	//	} else {
+	//		contextLogger.Fatalf("Error waiting for traces: got unexpected %v", event)
+	//	}
+	// }
+
+	var results gadgetv1alpha1.TraceList
+	start := time.Now()
+RetryLoop:
+	for {
+		results = gadgetv1alpha1.TraceList{}
+		err = traceRestClient.
+			Get().
+			Namespace("gadget").
+			Resource("traces").
+			VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+			Do(context.TODO()).
+			Into(&results)
+		if err != nil {
+			deleteTraces(contextLogger, traceRestClient, traceID)
+			contextLogger.Fatalf("Error getting traces: %q", err)
+		}
+
+		timeout := time.Now().Sub(start) > 2*time.Second
+		successNodeCount := 0
+		nodeErrors := make(map[string]string)
+		for _, i := range results.Items {
+			if i.Status.State == "Completed" || i.Status.State == "Started" {
+				successNodeCount++
+			} else {
+				if timeout {
+					nodeErrors[i.Spec.Node] = i.Status.OperationError
+					continue
+				}
+				time.Sleep(100 * time.Millisecond)
+				continue RetryLoop
+			}
+		}
+		for node, err := range nodeErrors {
+			contextLogger.Warningf("Error getting traces from node %q: %s", node, err)
+		}
+		if successNodeCount == 0 {
+			deleteTraces(contextLogger, traceRestClient, traceID)
+			contextLogger.Fatalf("Error getting traces from all nodes")
+		}
+		break RetryLoop
+	}
+
+	if callback == nil {
+		genericStreamsDisplay(contextLogger, client, params, &results)
+	} else {
+		callback(contextLogger, nodes, &results)
+	}
+	deleteTraces(contextLogger, traceRestClient, traceID)
+}
+
+func genericStreamsDisplay(contextLogger *log.Entry, client *kubernetes.Clientset, params *utils.CommonFlags, results *gadgetv1alpha1.TraceList) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	failure := make(chan string)
+
+	for _, i := range results.Items {
+		if params.Node != "" && i.Spec.Node != params.Node {
+			continue
+		}
+		go func(nodeName, namespace, name string) {
+			cmd := fmt.Sprintf("exec gadgettracermanager -call receive-stream -tracerid trace_%s_%s",
+				namespace, name)
+			var err error
+			err = execPod(client, nodeName, cmd, os.Stdout, os.Stderr)
+			if fmt.Sprintf("%s", err) != "command terminated with exit code 137" {
```

Comment:
Is there any way to avoid hardcoding `"command terminated with exit code 137"`. I think there are cases where the tool could fail with a different error.

Reply:
This test was originally written by @pothos in https://github.com/kinvolk/inspektor-gadget/commit/77e3e285927cd4eb18b2088bab249d90e4c97ee8 and is repeated a few times in different files.

It might be because it was written before Go 1.12 when it was not possible in Go to get the exit code. But since Go 1.12, it is possible to get the exit code without string manipulation:
```
    if exitError, ok := err.(*exec.ExitError); ok {
        return exitError.ExitCode()
    }
```

However, maybe we get the error message as a string from the API server, in that case we cannot avoid string manipulations.

Code 137 is 128 + 9(SIGKILL), meaning the process executed in the container was killed by that signal.

At the time this test was written, gadgets where killed with SIGKILL only, so it made sense to check for code 137:
```
err := execPodQuickStart(client, node.Name,
	fmt.Sprintf("sh -c \"touch /run/%s.pid; kill -9 \\$(cat /run/%s.pid); rm /run/%s.pid\"",
		tmpId, tmpId, tmpId))
```

But in the current code, bcc gadgets are stopped with SIGINT and then SIGKILL after a timeout by [bcc-wrapper.sh](https://github.com/kinvolk/inspektor-gadget/blob/main/gadget-container/gadgets/bcck8s/bcc-wrapper.sh#L89-L91), so it probably does not make sense anymore to hide the 137 error message:
```
    kill -SIGINT "$(cat $PIDFILE)" || true
    timeout 5s tail --pid="$(cat $PIDFILE)" -f /dev/null
    kill -9 "$(cat $PIDFILE)" || true
```

I think it could be clean up in a different PR.